File: tasking/sources/free_agent.2.f90
! @@name:	free_agent.2
! @@type:	F-free
! @@operation:	run
! @@expect:	success
! @@version:	omp_6.0

! export OMP_THREADS_RESERVE=”structured(64),free_agent(4)”

program main
  use omp_lib
  implicit none

  integer :: count = 0
  integer :: val

  val = fib(20)
  print *, "fib(20) =", val
  print *, "Number of tasks executed by free-agent threads =", count

contains

recursive function fib(n) result(res)
    integer, intent(in) :: n
    integer :: i, j, res

    if (n < 2) then
       res = n
       return
    end if

    ! Create tasks for recursive Fibonacci calculation
    !$omp task shared(i) threadset(omp_pool)
    i = fib(n - 1)
    !$omp end task

    !$omp task shared(j) threadset(omp_pool)
    j = fib(n - 2)
    !$omp end task

    !$omp taskwait

    ! Count tasks executed by free agent threads
    if (omp_is_free_agent()) then
       !$omp atomic
       count = count + 1
    end if

    res = i + j
end function fib

end program main
  

File: tasking/sources/free_agent.2.c
/*
* @@name:	free_agent.2
* @@type:	C
* @@operation:	run
* @@expect:	success
* @@version:	omp_6.0
*/

// export OMP_THREADS_RESERVE=”structured(64),free_agent(4)”

#include <stdio.h>
#include <omp.h>

int count = 0;
int fib(int n) {
   int i, j;

   if ( n<2 )
      return n;

   #pragma omp task shared(i) threadset(omp_pool)
      i=fib(n-1);
   #pragma omp task shared(j) threadset(omp_pool)
      j=fib(n-2);
   #pragma omp taskwait
   if ( omp_is_free_agent() ) {
      #pragma omp atomic
      count++;
   }
   return (i+j); 
}


int main(void) {
   int val;

   val = fib(20);
   printf(“fib(20) = %d\n”, val);
   printf(“Number of tasks executed by free-agent threads = %d\n”, count);

   return 0;
}
